File: ydg/인프런/코드/maze.java
package Inflearn;

//미로찾기
public class maze {
	
	/*
	 * 출구까지 나가는 법
	 *  >>> 현재 위치가 출구 or 이웃한 셀들 중 하나에서 현재 위치를 지나지 않고 출구까지 가는 경로가 있음
	 *  
	 *  1. BaseCase / RecursiveCase 생각
	 *      if (x,y)가 출구면 return true ----------------------------------- 1 : BaseCase
	 *      for : 아니면 각각 이웃한셀(x', y')에----------------------------- 2
	 *         if : 대해 갈 수 있으면---------------------------------------- 3
	 *              if :이웃한 셀이 출구면(x', y') return true--------------- 4 : RecursiveCase
	 *      아니면 return false; -------------------------------------------- 5 : BaseCase
	 *
	 *
	 *      if : 방문 or 벽이면 return false--------------------------------- 3 : BaseCase
	 *      if (x,y)가 출구면 return true ----------------------------------- 1 : BaseCase
	 *      for : 아니면 각각 이웃한셀(x', y')에----------------------------- 2
	 *              if :이웃한 셀이 출구면(x', y') return true--------------- 4 : RecursiveCase
	 *      아니면 return false; -------------------------------------------- 5 : BaseCase
	 *  
	 *  2. 무한루프에 빠지지 않게
	 *     : (x', y') 에서 다시 (x, y)로 오지 않게
	 */
	
	private static int N = 8;
	private static int [][] map = {
			{0,0,0,0,0,0,0,1},
			{0,1,1,0,1,1,0,1},
			{0,0,0,1,0,0,0,1},
			{0,1,0,0,1,1,0,0},
			{0,1,1,1,0,0,1,1},
			{0,1,0,0,0,1,0,1},
			{0,0,0,1,0,0,0,1},
			{0,1,1,1,0,1,0,0} 
	};
	
	private static final int PATHWAY = 0; //갈 수 있는 길
	private static final int WALL = 1;    //벽
	private static final int BLOCKED = 2; //방문한 길
	private static final int PATH = 3;    //지나온 길
	
	
	public static void main(String[] args) {
		System.out.println(findPath(0, 0));
	}
	
	private static boolean findPath(int x, int y) {
		if(x < 0 || x >= N || y < 0 || y >= N) {
			return false;
		}
		else if(map[x][y] != PATHWAY) {
			return false;
		}else if(x == N-1 && y == N-1){
			map[x][y] = PATH;
			return true; 
		}else {
			map[x][y] = PATH;
			if(findPath(x-1, y) || findPath(x, y+1) || findPath(x+1, y) || findPath(x, y-1) ) {
				return true;
			}
			map[x][y] =BLOCKED;
			return false;
		}
	}

}

File: gringrape/인프런 강의요약/code/maze.py
# boolean findPath(x,y)
# 	if (x,y) is not on the pathway or visited
# 		return false;
# 	else if (x,y) is the exit
# 		return true;
# 	else
# 		mark (x,y) as visited - Green
# 		for each neighboring cell (x',y') of (x,y)
# 			if findPath(x,y) is true
# 				return true;
#       change status of marking - Green -> Red
#      	return false;

# input data
N = 8
testMazeData = [
    [0, 0, 0, 0, 0, 0, 0, 1],
    [0, 1, 1, 0, 1, 1, 0, 1],
    [0, 0, 0, 1, 0, 0, 0, 1],
    [0, 1, 0, 0, 1, 1, 0, 0],
    [0, 1, 1, 1, 0, 0, 1, 1],
    [0, 1, 0, 0, 0, 1, 0, 1],
    [0, 0, 0, 1, 0, 0, 0, 1],
    [0, 1, 1, 1, 0, 1, 0, 0]
]

class Maze:
    mazeData = []
    N = 8
    START = [0, 0]
    EXIT = [N - 1, N - 1]
    
    # category of cells
    PATH = 0
    WALL = 1
    GREEN = 2
    RED = 3
    
    def __init__(self, mazeData):
        self.mazeData = mazeData
    def findPath(self, x, y):
        # Out of range 
        if x < 0 or x >= self.N or y < 0 or y >= self.N:
            return False
        
        # Not on the Path or Visited
        elif self.mazeData[x][y] == self.WALL or self.mazeData[x][y] == self.GREEN or self.mazeData[x][y] == self.RED:
            return False
        
        # BASE CASE - EXIT
        elif self.EXIT == [x, y]:
            return True
        
        # RECURSIVE CASE
        else:
            self.mazeData[x][y] = self.GREEN
            if self.findPath(x, y - 1) or self.findPath(x, y + 1) or self.findPath(x - 1, y) or self.findPath(x + 1, y):
                return True
            self.mazeData[x][y] = self.RED
            return False

## Tests

# # 1. Maze setting maze data test -> o.k
# maze = Maze(testMazeData)
# for row in maze.mazeData:
#     print(row)
# print(maze.START, maze.EXIT)

# # 2. find path -> fail -> success
# maze = Maze(testMazeData)
# maze.findPath(0, 0)
# for row in maze.mazeData:
#     print(row)

# # 3. i in range(-1, 2, 2) -> success
# for i in range(-1, 2, 2):
#     print(i)

# test case 2의 오류 원인 분석
# - 주된 원인: 이미 방문한 셀을 표시해주는 기능이 동작하지 않아서, 이미 방문한 셀을 다시 호출하는 무한 loop 에 빠졌다.
#   - 이미 방문한 셀을 표시해주는 기능이 작동하지 않은 이유
#       - curr = self.mazeData[x][y]
#       - curr의 값을 바꾸었을때 mazeData[x][y] 의 값이 바뀌지는 않았다.  
#       - 복사되는 것이 reference가 아닌 value 이기 때문에 그렇다. 
#   - 해결책: reference 변수에 복사해야 한다. 